kratos/elements/distance_calculation_flux_based_element.cpp

```diff
@@ -0,0 +1,502 @@
+        avg_DN_DX += DN_DX_container[i] /static_cast<double>(TNumNodes);
+    }
+    
+    //adjusting RHS to get |grad|=1
+    if(rCurrentProcessInfo[NL_ITERATION_NUMBER] > 1) {
+        const array_1d<double, TDim> avg_grad = prod(trans(avg_DN_DX), nodal_values);
+        const double grad_modulus = norm_2(avg_grad);
+
+        const double new_correction_coeff = 1.0/grad_modulus;
+        const double relaxation_factor = 0.5;
+        mCorrectionCoefficient = mCorrectionCoefficient *(1.0-relaxation_factor) + new_correction_coeff*relaxation_factor;
+    }
+    else {
+        mCorrectionCoefficient = 1.0;
+    }
+
+    //computing element size(for Tau)
+    const double h = ComputeH(avg_DN_DX);
+
+    //DECIDE IF CONVECTION IS ACTIVE/INACTIVE
+   //we decide by checking if flow converges to /diverges from this element
+   //if all velocity vectors are oriented in the same direction, then convection is valid
+   //but if the element acts as a "sink" or "source", convection is turned off to avoid instabiliities.
+    bool add_convection = true;
+    double average_unit_vel_misaligment = 0.0;
+    for(unsigned int i=0; i<NumNodes; i++) {
+        average_unit_vel_misaligment += norm_2( v_unit[i]  - avg_v_unit );
+    }
+    average_unit_vel_misaligment/=static_cast<double>(NumNodes);
+    if(average_unit_vel_misaligment>0.33) add_convection=false;
+
+    //checking if the flow comes from face that has neighbour element (only if it is not inlet)
+    if( !has_fixed_node && this->Has(NEIGHBOUR_ELEMENTS) && this->GetValue(NEIGHBOUR_ELEMENTS).size()==GetGeometry().FacesNumber())
+    {
+        double value_lowest_prod = 0.0;
+        unsigned int face_lowest_prod = 0;
+        const auto elem_boundaries = GetGeometry().LocalSpaceDimension()==3 ? GetGeometry().GenerateFaces() : GetGeometry().GenerateEdges() ;
+
+        for(unsigned int i=0; i<elem_boundaries.size(); i++) {
+            Point cond_center = elem_boundaries[i].Center();
+            const array_1d<double,3> unit_normal = elem_boundaries[i].UnitNormal(cond_center);
+            const double face_result = inner_prod(avg_v_unit,unit_normal);
+            if(face_result<value_lowest_prod){
+                value_lowest_prod = face_result;
+                face_lowest_prod=i;
+            }
+        }
+        if( this->GetValue(NEIGHBOUR_ELEMENTS)(face_lowest_prod).get()==nullptr){
+            add_convection=false;
+        }
+    }
+
+    //TERMS TO BE ASSEMBLED
+    //terms which multiply the gradient of phi
+    BoundedMatrix<double,NumNodes, NumNodes> aux2 = ZeroMatrix(NumNodes, NumNodes); //terms multiplying phi
+    
+    //source term
+    array_1d<double, NumNodes> rhs_volumetric_heat = ZeroVector(NumNodes);
+
+    //Gauss point container (using nodal integration)
+    const bounded_matrix<double, NumNodes, NumNodes> Ncontainer=IdentityMatrix(NumNodes); 
+
+    //Looping Gauss Points
+    for (unsigned int igauss = 0; igauss<NumNodes; igauss++) {
```

Comment:
```suggestion
    for (unsigned int igauss = 0; igauss<n_gauss; igauss++) {
```
Same comment, I wouldn't assume this just in case we change the uadrature.

Reply:
ok